R/create_single_dose_dataset.R

```diff
@@ -0,0 +1,272 @@
+    CDISCSubmissionValue %in% c("EVERY WEEK") ~ "WEEK",
+    CDISCSubmissionValue %in% c("BID", "TID", "QAM", "QPM", "QHS",
+                                "QD", "QN", "QID", "QOD") ~ "DAY",
+    CDISCSubmissionValue %in% c("QM", "BIM") ~ "MONTH",
+    CDISCSubmissionValue == "PA" ~ "YEAR",
+  )) %>%
+mutate(
+  DayConversionFactor = case_when(
+    DoseWindow == "DAY" ~ 1,
+    DoseWindow == "WEEK" ~ (1 / 7),
+    DoseWindow == "MONTH" ~ (1 / 30.4375),
+    DoseWindow == "YEAR" ~ (1 / 365.25),
+  )
+)
+
+#' Create dataset of single doses
+#'
+#' Derives dataset of single dose from aggregate dose information
+#'
+#' @param dataset Input dataset
+#'
+#'   The columns specified by `by_vars`, `dose_freq`, `start_date` and the `end_date`
+#'   parameters are expected.
+#'
+#' @param by_vars Variables to group by (created by `dplyr::vars`).
+#'
+#'   These variables are used to identify unique rows for each aggregate dose.
+#'
+#'   Example: vars(STUDYID, USUBJID, EXTRT)
+#'
+#' @param dose_freq The dose frequency
+#'
+#'   The aggregate dosing frequency used for multiple doses in a row.
+#'
+#'   Default: EXDOSFRQ
+#'
+#'   Permitted Values: defined by lookup table.
+#'
+#' @param start_date The start date
+#'
+#'   A date or date-time object is expected.
+#'
+#'   Default: `ASTDT`
+#'
+#' @param end_date The end date
+#'
+#'   A date or date-time object is expected.
+#'
+#'   Default: `AENDT`
+#'
+#' @param lookup_table The dose frequency value lookup table
+#'
+#'   The table used to look up `dose_freq` values and determine the appropriate
+#'   multiplier to be used for row generation. If a lookup table other than the
+#'   default is used, it must have columns `DoseWindow`, `DoseCount`, and
+#'   `DayConversionFactor`
+#'
+#'   Default: `dose_freq_lookup`
+#'
+#'  Permitted Values for `DoseWindow`: "DAY", "WEEK", "MONTH", "YEAR"
+#'
+#' @param lookup_column The dose frequency value column in the lookup table
+#'
+#'   The column of `lookup_table`.
+#'
+#'   Default: `CDISCSubmissionValue` (column of `dose_freq_lookup`)
+#'
+#'
+#' @details Each aggregate dose row is split into multiple rows which each represent
+#' a single dose.The number of completed dose periods between `start_date` and
+#' `end_date` is calculated with `compute_duration` and multiplied by the count
+#' of doses per period. `start_date` and `end_date` are then adjusted by an
+#' appropriate increment using `DayConversionFactor`.
+#'
+#' @author Michael Thorpe, Andrew Smith
+#'
+#' @return The input dataset with a single dose per row.
+#'
+#' @export
+#'
+#' @examples
+#' data <- tibble::tribble(
+#'   ~USUBJID, ~EXDOSFRQ, ~ASTDT, ~AENDT,
+#'   "P01", "Q2D", lubridate::ymd("2021-01-01"), lubridate::ymd("2021-01-07"),
+#'   "P01", "Q3D", lubridate::ymd("2021-01-08"), lubridate::ymd("2021-01-15"),
+#'   "P01", "EVERY 2 WEEKS", lubridate::ymd("2021-01-15"), lubridate::ymd("2021-01-29")
+#' )
+#'
+#' create_single_dose_dataset(data, by_vars = vars(USUBJID))
```

Comment:
An example with custom lookup table would be helpful.

Reply:
Could you add comment before the second example to separate them?

Reply:
are the lubridate calls necessary?  short of clutters up the example

Reply:
You need to add `library(lubridate)` like in https://pharmaverse.github.io/admiral/reference/derive_vars_dy.html#examples.